### DISCUSSION TITLE
YOLOv5 losses questions

### DISCUSSION THREAD

**Discussion by antimo22** [ultralytics/yolov5#6276] [Q&A]
Hi everybody, I have some doubt about the losses used in this version. I've already tried to understand something from the loss.py script, but I wasn't able.

First of all, I've found this article: https://www.nature.com/articles/s41598-021-01084-x
Where the author describes the loss in this way:

![issue_1](https://user-images.githubusercontent.com/65121908/149117819-c7f75c3d-c4c3-421e-a56d-1171d0ec9764.png)
![coiuloss](https://user-images.githubusercontent.com/65121908/149117823-a7a7bc2e-afba-449a-92f7-c13b8c236cfd.png)

Now, looking in the issues, I noticed that the classification loss term is correct, except for the second sum, which is on the classes.
Also, the term called Lciou should be correct, and corresponds to the objectness loss.
My doubts are about the other term, here called confidence loss, that in v5 is called box loss. Looking at the loss.py file, I don't think that these terms matches. Also, I've read in the issues that this term is slightly different from the relative term in the previous yolo versions:

![box_loss](https://user-images.githubusercontent.com/65121908/149125975-48438dcc-5880-4d3f-8b99-fe8189b5ed48.png)

So I would be grateful for clarification regarding this term. Thanks in advice.

**Comment by antimo22** [comment]
Hey, maybe this article has the answers to my questions: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8610656/

![loss_def](https://user-images.githubusercontent.com/65121908/149151239-62c536f6-4c7c-46eb-bd68-d21c5175023c.png)

Please let me know if it's all correct. Thanks.

**Comment by joihn** [comment]
I agree, I'm also struggling to derivate the mathematical expression for the loss, 
it seems the 2 cited paper do not match with current code implementation.

**Comment by lucysumi** [comment]
Could you please clarify which is the correct formula for Lbox @glenn-jocher as different links and different papers says different things about Yolov5 loss function?
Would be very grateful if anyone could help. Thanks !

**Reply by glenn-jocher** [reply]
@lucysumi losses are defined in utils/loss.py

**Comment by MZeinab** [comment]
@glenn-jocher
Hi, could you please write the exact formula that you used in loss.py, I don't understand what exactly it is, 
it will be so useful for us and even for adding it to our articles. I checked the loss.py, if I understood correctly you mean this:

`   def __call__(self, p, targets):  # predictions, targets
        lcls = torch.zeros(1, device=self.device)  # class loss
        lbox = torch.zeros(1, device=self.device)  # box loss
        lobj = torch.zeros(1, device=self.device)  # object loss
        tcls, tbox, indices, anchors = self.build_targets(p, targets)  # targets

        # Losses
        for i, pi in enumerate(p):  # layer index, layer predictions
            b, a, gj, gi = indices[i]  # image, anchor, gridy, gridx
            tobj = torch.zeros(pi.shape[:4], dtype=pi.dtype, device=self.device)  # target obj

            n = b.shape[0]  # number of targets
            if n:
                # pxy, pwh, _, pcls = pi[b, a, gj, gi].tensor_split((2, 4, 5), dim=1)  # faster, requires torch 1.8.0
                pxy, pwh, _, pcls = pi[b, a, gj, gi].split((2, 2, 1, self.nc), 1)  # target-subset of predictions

                # Regression
                pxy = pxy.sigmoid() * 2 - 0.5
                pwh = (pwh.sigmoid() * 2) ** 2 * anchors[i]
                pbox = torch.cat((pxy, pwh), 1)  # predicted box
                iou = bbox_iou(pbox, tbox[i], CIoU=True).squeeze()  # iou(prediction, target)
                lbox += (1.0 - iou).mean()  # iou loss

                # Objectness
                iou = iou.detach().clamp(0).type(tobj.dtype)
                if self.sort_obj_iou:
                    j = iou.argsort()
                    b, a, gj, gi, iou = b[j], a[j], gj[j], gi[j], iou[j]
                if self.gr < 1:
                    iou = (1.0 - self.gr) + self.gr * iou
                tobj[b, a, gj, gi] = iou  # iou ratio

                # Classification
                if self.nc > 1:  # cls loss (only if multiple classes)
                    t = torch.full_like(pcls, self.cn, device=self.device)  # targets
                    t[range(n), tcls[i]] = self.cp
                    lcls += self.BCEcls(pcls, t)  # BCE

                # Append targets to text file
                # with open('targets.txt', 'a') as file:
                #     [file.write('%11.5g ' * 4 % tuple(x) + '\n') for x in torch.cat((txy[i], twh[i]), 1)]

            obji = self.BCEobj(pi[..., 4], tobj)
            lobj += obji * self.balance[i]  # obj loss
            if self.autobalance:
                self.balance[i] = self.balance[i] * 0.9999 + 0.0001 / obji.detach().item()

        if self.autobalance:
            self.balance = [x / self.balance[self.ssi] for x in self.balance]
        lbox *= self.hyp['box']
        lobj *= self.hyp['obj']
        lcls *= self.hyp['cls']
        bs = tobj.shape[0]  # batch size

        return (lbox + lobj + lcls) * bs, torch.cat((lbox, lobj, lcls)).detach()`
 



I trained my custom model with your friendly tutorials, yolov5 v6.0, I want to write a report about it and explain the exact loss function but I am really confused. I'll appreciate it if you could guide
Best of all,

**Comment by joihn** [comment]
https://docs.ultralytics.com/yolov5/tutorials/architecture_description/#4-additional-features

**Reply by ghost** [reply]
But this document is still outdated and does not cover the seg loss of YOLOv5-v7.0.

**Comment by venkanna37** [comment]
I think this paper mentioned three losses properly that used in YOLOv8. Here is the paper link: [Real-Time Flying Object Detection with YOLOv8](https://arxiv.org/pdf/2305.09972.pdf)
